addons/website_sale_collect/controllers/delivery.py

```diff
@@ -10,27 +10,40 @@ class InStoreDelivery(Delivery):
     @route()
     def website_sale_get_pickup_locations(self, zip_code=None, **kwargs):
         """ Override of `website_sale` to set the pickup in store delivery method on the order in
-        order to retrieve pickup locations when called from the the product page.
+        order to retrieve pickup locations when called from the product page. If there is no order
+        create a temporary one to display pickup locations.
         """
         if kwargs.get('product_id'):  # Called from the product page.
-            order_sudo = request.website.sale_get_order(force_create=True)
+            order_sudo = request.website.sale_get_order()
             in_store_dm = request.website.sudo().in_store_dm_id
-            if order_sudo.carrier_id.delivery_type != 'in_store':
+            if not order_sudo:  # Pickup location requested without a cart creation.
+                temp_order_sudo = self._create_temp_order(carrier_id=in_store_dm)
+                return temp_order_sudo._get_pickup_locations(zip_code, **kwargs)  # Skip super
+            elif order_sudo.carrier_id.delivery_type != 'in_store':
                 order_sudo.set_delivery_line(in_store_dm, in_store_dm.product_id.list_price)
         return super().website_sale_get_pickup_locations(zip_code, **kwargs)
 
     @route('/shop/set_click_and_collect_location', type='jsonrpc', auth='public', website=True)
     def shop_set_click_and_collect_location(self, pickup_location_data):
-        """ Set the pickup location and the in-store delivery method on the current order.
+        """ Set the pickup location and the in-store delivery method on the current order or created
```

Comment:
Do we really need to force the order to be created here?
Since this route is only called from the checkout page, you're guaranteed to have a cart.
I believe we should return early if there is no order instead of creating one :thinking:

Reply:
this is called when the user chooses a location from the location selector in /product page. Maybe he can first choose a location and then add a product.

Reply:
I guess I misunderstood the comment below then

>         This route is distinct from /website_sale/get_pickup_locations as the latter is only called
        from the checkout page after the delivery method is selected.

Reply:
Will try to make it clearer!

Reply:
Thanks, it's better already.
I believe we could improve it even more someday and store the pickup location data in the session until the client really adds a product to their cart, but it's already better :+1: